libenv/sysinfo.c

```diff
@@ -3465,6 +3465,69 @@ static void SysOSNameHuman(EvalContext *ctx)
 
 /*****************************************************************************/
 
+void SysOsVersionMajor(EvalContext *ctx)
+{
+    const char *flavor_const = EvalContextVariableGetSpecialString(
+        ctx, SPECIAL_SCOPE_SYS, "flavor");
+    char *flavor = SafeStringDuplicate(flavor_const);
+
+    if (flavor == NULL)
+    {
+        EvalContextVariablePutSpecial(ctx, SPECIAL_SCOPE_SYS,
+                                      "os_version_major", "Unknown",
+                                      CF_DATA_TYPE_STRING, "source=agent");
+        return;
+    }
+
+    char *major = NULL;
+    char *minor = NULL;
+    for (char *sp = flavor; *sp != '\0'; sp++)
+    {
+        if (major == NULL && isdigit(*sp))
+        {
+            major = sp;
+        }
+        else if (minor == NULL && isdigit(*sp))
```

Comment:
If major version has mutiple digits, minor will be assigned to the second digit of major.

In general, I think something like this is less error prone:

```C
while (*ptr != '\0' && !isdigit(*ptr))
{
    ++ptr; // Looking for first digit
}
if (*ptr == '\0')
{
    // Return early without assigning major
    // Or perhaps add an assert / abort / error here
    return;
}
major = ptr;
while (isdigit(*ptr))
{
    ++ptr; // Looking for end of first version number
}
if (*ptr == '.')
{
    *ptr = '\0';
    ptr++;
    if (isdigit(*ptr)) {
        minor = ptr;
        while (isdigit(*ptr)) ++ptr; // Looking for end of second version number (minor)
        *ptr = '\0':
    }
}
else
{
    // Assert that it ended at a zero byte, or do something else?
}
```

Reply:
Good catch. Indeed it's safer to do this step by step.